django/core/validators.py

```diff
@@ -86,7 +86,14 @@ class URLValidator(RegexValidator):
     hostname_re = r'[a-z' + ul + r'0-9](?:[a-z' + ul + r'0-9-]{0,61}[a-z' + ul + r'0-9])?'
     # Max length for domain name labels is 63 characters per RFC 1034 sec. 3.1
     domain_re = r'(?:\.(?!-)[a-z' + ul + r'0-9-]{1,63}(?<!-))*'
-    tld_re = r'\.(?:[a-z' + ul + r']{2,63}|xn--[a-z0-9]{1,59})\.?'
+    tld_re = (
+        '\.'                                # dot
+        '(?!-)'                             # can't start with a dash
+        '(?:[a-z-' + ul + ']{2,63}'         # domain label
```

Comment:
Forgot to mention earlier,  but on first look I found `[a-z-' + ul` a little confusing because of the dash between two ranges that actually serves as a dash and not a range separator. I think it would be more readable as `[a-z' + ul + r'-]` (similar to how it is in `domain_re` above).